localeredirector/services/LocaleRedirectorService.php

```diff
@@ -74,15 +74,14 @@ public function getBrowserLanguageMatch()
   private function newUrl($locale)
   {
     $qs = $this->querystring ? '?' . $this->querystring : '';
-    $segments = craft()->request->segments;
-    $lastSegment = end($segments);
-    $criteria = craft()->elements->getCriteria(ElementType::Entry);
-    $criteria->slug = $lastSegment;
-    $entry = $criteria->first();
+    $element = craft()->urlManager->getMatchedElement();
 
-    if (!empty($entry)) {
-      $new = craft()->elements->getElementById($entry->id, ElementType::Entry, $locale);
-      return UrlHelper::getSiteUrl($new->uri, null, null, $locale) . $qs;
+    if ($element) {
+      $newElement = craft()->elements->getElementById($element->id, ElementType::Entry, $locale);
+      $isHome = $newElement->uri == '__home__';
+      $path = !$isHome ? $newElement->uri : '';
```

Comment:
Had to add a small test for the homepage uri. Because of the change above, homepage redirects were going to `/es/__home__`.